File: 8.interpolation search.c
//write a program to implement interpolation search over an array
#include<stdio.h>
#include<stdbool.h>
int main(){
    int a,val;
    printf("enter the number of elements: \n");
    scanf("%d",&a);
    int arr[a];
    printf("enter sorted elements :\n");
    for(int i = 0;i<a;i++){
        scanf("%d",&arr[i]);
    }
    printf("enter element to search: \n");
    scanf("%d",&val);
    int low = 0,high = a-1;
    bool isfound = false;
    int mid;

    while(arr[low]<=val && low<=high && val<=arr[high]){
        mid = low+(((val-arr[low])*(high-low))/(arr[high]-arr[low]));
        if(val == arr[mid]){
            isfound = true;
            break;
        }
        else if(arr[mid]>val){
            high = mid-1;
        }
        else{
            low = mid+1;
        }
    }
    if(!isfound){
        printf("the element is not found\n");
    }
    else{
        printf("the element is found at position %d",mid+1);
    }
}

/*
enter the number of elements:
5
enter sorted elements :
12
23
34
45
56
enter element to search:
56
the element is found at position 5


enter the number of elements:
5
enter sorted elements :
12
23
34
45
56
enter element to search:
18
the element is not found


*/

File: 8.interpolation search.py
# write a program to implement interpolation search over an array
import numpy as np

a = int(input("enter the number of elements: "))
arr = np.zeros(a,int)
print("enter sorted elements: ")
for i in range(0,a):
    arr[i] = int(input())

val = int(input("enter element to search "))

low = 0;high = a-1
isfound = False
mid = 0

while(arr[low]<=val and low<=high and val<=arr[high]):
    mid = int(low+(((val-arr[low])*(high-low))/(arr[high]-arr[low])))
    if(val == arr[mid]):
        isfound = True
        break
    elif(arr[mid]>val):
        high = mid-1
    else:
        low = mid+1
if(not isfound):
    print("the element is not found")
else:
    print("the element is found at position",mid+1)
   